Feature coalescing

webmiche opened the pull request:
Updated version of the PR after rebasing

webmiche committed:
ported ineqType

webmiche committed:
added files with libint text

webmiche committed:
ported more functionality

webmiche committed:
[Coalesce] contained case

webmiche committed:
[Porting] testcase, functionality but protrusion

webmiche committed:
[Porting] finished porting

webmiche committed:
[Bugfix] fixed all remaining bugs

webmiche committed:
[Refactoring] Info struct

webmiche committed:
[Refactoring] reduced infostruct

webmiche committed:
[Refactoring] changed comments

webmiche committed:
[Refactoring] added ArrayRefs

webmiche committed:
[Refactoring] implemented Feedback Pt.1

webmiche committed:
[Refactoring] fully implemented feedback

webmiche committed:
[Refactoring] implemented feedback Pt2

webmiche committed:
[Refactoring] implemented feedback pt3

webmiche committed:
[Refactoring] implemented feedback, pt 4

webmiche committed:
[Refactoring] changed dump to const

webmiche committed:
[Refactoring] added documentation

webmiche committed:
[Refactoring] changed test vector initialization

webmiche committed:
[Refcatoring] added examples to comments

webmiche committed:
[Refactoring] added documentation

webmiche committed:
[Refactoring] added documentation

webmiche committed:
[Refactoring] decoupled adjEq Cases pure/non pure, added docu

webmiche committed:
[Refactoring] added docu

webmiche committed:
[Refactoring] added dump as struct method

superty commented on mlir/include/mlir/Analysis/Presburger/Coalesce.h:
```diff
@@ -0,0 +1,41 @@
+// coalesces a set according to the "integer set coalescing" by sven
+// verdoolaege.
+//
+// Coalescing task two convex BasicSets and tries to figure out, whether the
```
nit: "Coalescing _takes_..."?

webmiche replied:
yes, thx

superty commented on mlir/include/mlir/Analysis/Presburger/Coalesce.h:
```diff
@@ -0,0 +1,41 @@
+
+void dump(const ArrayRef<int64_t> cons);
+
+// compare two constraints and gives true, even if they are stretched
```
Please elaborate here on what "stretched" means.

webmiche replied:
I rewrote this to:
compare two constraints and give true if they are equal. Can also handle cases in which for some integers a and b, c1 = a/b * c2. This is the same constraint but stretched, which doesn't influence it's hyperplane.

superty commented on mlir/include/mlir/Analysis/Presburger/Coalesce.h:
```diff
@@ -0,0 +1,41 @@
+bool sameConstraint(ArrayRef<int64_t> c1, ArrayRef<int64_t> c2);
+
+// tries to find a constraint, that is a linear combination of valid and
+// invalid, and touches bs/is REDUNDANT in the most limited way possible
```
Please elaborate here on what "most limited way possible" means.

webmiche replied:
I rewrote this to:
rotates invalid around valid, until it becomes redundant. It does this by adding the smallest multiple of valid to invalid, such that the result is redundant.

superty commented on mlir/include/mlir/Analysis/Presburger/Coalesce.h:
```diff
@@ -0,0 +1,41 @@
+
+// takes a BasicSet bs, a constraint ineq of that basicSet and the vector cut of
+// constraints, that were typed as cutting bs. Computes wether the part of ineq,
+// that lies within bs, is redundant for all constraints of cut
```
nit: Should this be, "compute whether _all constraints of cut are redundant_ for the part of ineq that lies within bs"?

superty replied:
Please write that by "the part of ineq that lies within bs", it means the part of bs that satisfies ineq with equality, or something like that, if that is what it is supposed to do.

webmiche replied:
Yes, that's what it is supposed to do. I rewrote this.

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+struct Info {
+  SmallVector<ArrayRef<int64_t>, 8> redundant;
+  SmallVector<ArrayRef<int64_t>, 8> cut;
+  Optional<ArrayRef<int64_t>> adj_ineq;
```
nit: LLVM uses camelCase for variables, not snake_case.

webmiche replied:
rewrote everything, _ only occurs in types and vector instructons now.
And in the enumeration of IneqType. We might have to change it there too (ADJ_INEQ --> ADJINEQ)

webmiche replied:
I just saw your comment below, nevermind the last line ;)

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+void shift(SmallVectorImpl<int64_t> &t, int amount);
+
+// add eq as two inequalities to target
+void addAsIneq(const ArrayRef<ArrayRef<int64_t>> eq,
```
Maybe indicate what is being added in the function name, e.g., `addEqualitiesAsInequalities`.

webmiche replied:
done

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+//
+// returns true if it has neither encountered a separate constraint nor more
+// than one adjacent inequality
+bool classify_ineq(Simplex &simp,
```
LLVM style would have this be `classifyIneq`

webmiche replied:
done

superty commented on mlir/include/mlir/Analysis/Presburger/Simplex.h:
```diff
@@ -128,6 +128,8 @@ class Simplex {
 public:
   enum class Direction { Up, Down };
 
+  enum class IneqType { REDUNDANT, SEPARATE, CUT, ADJ_EQ, ADJ_INEQ };
```
LLVM style would use UpperCamelCase for enum values

webmiche replied:
made ADJ_INEQ --> ADJINEQ and ADJ_EQ --> ADJ_EQ

Dinistro replied:
I think the idea was to change ADJ_INEQ to AdjIneq

webmiche replied:
Ah okay, I thought generally enumerations are done in UPPERCASE, which doesn't seem to be the case as Direction is UpperCameCase now too. I will change ADJINEQ -> AdjIneq and ADJEQ -> AdjEq.

webmiche replied:
I just realiesed that I will have to do the same for the other Types two '^^

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+// classifies all constraints into redundant, cut, adj_ineq or t, where t stands
+// for a constraint adjacent to a the other polytope
+//
+// returns true if it has not encountered a separate constraint or more than one
```
A little explanation of why this is the return value would be nice.

webmiche replied:
I added:
returns true if it has neither encountered a separate constraint nor more than one adjacent inequality. This has a similar idea as short-circuiting behind it. The moment either a separate or two adjacent constraints are encountered, the polytope cannot be coalesced anymore, so we can move to the next tuple.

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+//
+// returns true if it has neither encountered a separate constraint nor more
+// than one adjacent inequality
+bool classify_ineq(Simplex &simp,
```
A little explanation of why this is the return value would be nice. Also below for `classify`

webmiche replied:
I added (same as below):
returns true if it has neither encountered a separate constraint nor more than one adjacent inequality. This has a similar idea as short-circuiting behind it. The moment either a separate or two adjacent constraints are encountered, the polytope cannot be coalesced anymore, so we can move to the next tuple.

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+}
+
+void addAsIneq(ArrayRef<ArrayRef<int64_t>> eq,
+               SmallVectorImpl<ArrayRef<int64_t>> &target) {
```
Would it be possible to use a `MutableArrayRef` of `SmallVector`s, and avoid the pointer stuff below?

webmiche replied:
The pointer stuff is needed, to make all constraints become `ArrayRef`s, which suffices for the rest of the program and AFAIU is much faster than using something that is mutable like `MutableArrayRef`or something similar.

I didn't actually give much thought into swapping `SmallVectorImpl` to `MutableArrayRef` as I felt like we will need to redo a lot of those things as soon as the Constraint class and the BasicSet class have landed.

I added a general TODO: look at `MutableArrayRef` to the code

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+    target.push_back(curr);
+    // TODO: fix this memory leak
+    SmallVector<int64_t, 8> *inverted = new SmallVector<int64_t, 8>();
+    for (size_t j = 0; j < curr.size(); j++) {
```
range-based for loop?

webmiche replied:
done

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+
+void addAsIneq(ArrayRef<ArrayRef<int64_t>> eq,
+               SmallVectorImpl<ArrayRef<int64_t>> &target) {
+  for (size_t i = 0; i < eq.size(); i++) {
```
range-based for loop?

webmiche replied:
done

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+                          const SmallVector<ArrayRef<int64_t>, 8> &cut) {
+  Simplex simp(bs);
+  simp.addEquality(ineq);
+  for (size_t i = 0; i < cut.size(); i++) {
```
range-based for loop?

webmiche replied:
done

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+}
+
+PresburgerSet mlir::coalesce(PresburgerSet &set) {
+  PresburgerSet new_set(set.getNumDims(), set.getNumSyms());
```
LLVM uses camelCase instead of snake_case

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+
+// compute the cut case and return whether it has worked.
+//
+// The cut case is the case, for which a polytope only has REDUNDANT and CUT
```
As far as I understand, the sloped cut constraint in the left diagram is not totally contained in the other polytope? What am I missing? Please elaborate in the documentation.

webmiche replied:
The sloped cut constraint actually is contained within the left polytope. Note that due to them overlapping, I coudn't fully draw the left polytope. I think this is a limitation of the kind of representation and I will try to come up with a better example that is not limited like that.

webmiche replied:
I changed the example, hopefully, it is clearer now. I tried to do the same example as the integer set coalescing paper uses for the cut case.

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+//   ____          ____
+//  |    \        |    \
+//  |     \   ==> |     \
+//  |_____|_      |______\
```
The second polytope is not clearly visible in this diagram. Please write a line pointing out where it is.

webmiche replied:
I added:
In the example below, Polytope A is the big one on the left, Polytope B is the small line at the right side. that gets included after coalescing.

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+      return false;
+    }
+    auto res = result.getValue();
+    if (res.num <= -2 * res.den) {
```
Might be cleaner to write it as `res <= Fraction(-2, 1)`.

webmiche replied:
done

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+}
+
+// helperfuncton to convert arrayRefs to SmallVectors
+static SmallVector<int64_t, 8> arrayRefToSmallVector(ArrayRef<int64_t> ref) {
```
You can use `llvm::to_vector`.

webmiche replied:
How do you mean this? I tried substituting this fucntion with llvm::to_vector, but this seems to take a size parameter and not an `ArrayRef.`

Dinistro replied:
`to_vector` takes a `Range`. Note that this isn't per se a specific class as they extract begin and end with `std::begin` and `std::end`. 

https://llvm.org/doxygen/SmallVector_8h_source.html line 964

It seems like it also requires a `Size` template parameter that in you case would be 8.

webmiche replied:
I understood this now and use `llvm::to_vector`for this.

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+bool mlir::sameConstraint(ArrayRef<int64_t> c1, ArrayRef<int64_t> c2) {
+  Fraction ratio(0, 1);
+  if (c1.size() != c1.size()) {
+    return false;
```
Should this rather be an assertion failure? Can this ever happen normally?

webmiche replied:
This cannot happen within my code, but as I define this for the whole mlir namespace, I decided to cover this case.

I will change this to an assertion, as someone who is actually comparing constraints with different sizes might be doing something else wrong.

Note: Depending on our representation of existential dimensions and int divisions, this case might happen actually AFAIU.

superty commented on mlir/lib/Analysis/Presburger/Coalesce.cpp:
```diff
@@ -0,0 +1,909 @@
+    if (c2[i] != 0 && ratio.num == 0) {
+      ratio.num = c1[i];
+      ratio.den = c2[i];
+      if (ratio.den * c1[i] != ratio.num * c2[i]) {
```
This check is redundant, right?

webmiche replied:
agreed and done

superty reviewed (commented):
Thanks for this PR. Some general comments:

- Especially since we plan to upstream this eventually, we should follow LLVM style, and have sufficiently detailed documentation, so I have left several comments to that effect.

- Is it possible to change the const SmallVector references to ArrayRefs? The SmallVectors of ArrayRefs included.

- Try to use range-based for loops when possible.

- Use camelCase instead of snake_case for variable and function names.

webmiche committed:
[Refactoring] implemented feedback from 10.07

webmiche committed:
[Refactoring] changed IneqType to UpperCamelCase

webmiche committed:
[Refactoring] changed all of IneqType enum

webmiche committed:
[Refactoring] changed for loops to range-based

webmiche committed:
[Refactoring] changed size_t and comments

webmiche committed:
[Bug] found bug, added TODO's for it

webmiche committed:
[Tests] added performance testcases

webmiche committed:
[feature] add coalesce operation to dialect

webmiche committed:
[Bug] hack for memory leak

webmiche committed:
[Refactoring] introduce llvm::to_vector

webmiche committed:
[Testing] added test cases and generator

webmiche committed:
[Feature] add verify(coalesceOp), producesPresburgerSet trait

webmiche committed:
[Bug] fix issue with constant in generated sets

webmiche committed:
[Bug] turn equality check off, fix testcases

webmiche committed:
[Refactoring] add dimensionality assertions

webmiche committed:
[Bug] turn on equality checking

Dinistro committed:
feature: apply operation for PresburgerExpr

superty committed:
[Simplex::ineqType] support AdjIneq

Dinistro committed:
feature: presburger.apply lowering prototype

superty committed:
[Simplex::separationType] fix adjIneq for non-redundant constraints

Dinistro committed:
testing: add more expression lowering tests

Dinistro committed:
refactoring: Small optimization + renaming

Dinistro committed:
feature: Presburger evaluate pass

Moved all canonicalization passes, i.e. set operation evaluations, into
a Presburger evaluate pass. It can be triggered with
`-presburger-evaluate`

Dinistro committed:
feature: Commutative trait for Presburger ops

Dinistro committed:
feature: Presburger canonicalization pass

Dinistro committed:
fix: remove SetAttr equality check

Dinistro committed:
fix: tests that used equal attribute detection

superty committed:
[Set::triviallyEqual] add trivially equal heuristic for Sets

Dinistro committed:
feature: remove redundant type in equal operation

superty committed:
Revert "[Set::triviallyEqual] add trivially equal heuristic for Sets"

This reverts commit 390bb18e7581cdc129cabce1ea9015bec3e389ed.

webmiche committed:
[Bug] generate new test cases, add failing gtest

superty committed:
Add regression test for 00f08fd86b2b95828e78819acea5a07f81383569

Dinistro committed:
fix: union of empty sets now works properly

webmiche committed:
[Bug] fix Parser instantiation
